### ISSUE TITLE
Allow specifying DNS resolvers for Let's Encrypt DNS challenge

### ISSUE THREAD

**Issue by jaen** [containous/traefik#3585]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

I want to use Træfik in a setup where there is a certain external domain `registry.developer.com` (that only client's network can access from the wide web) and a certain internal `registry.int.developer.com` domain (that can be accessed in our network, behind VPN) that point to the same service (say, a Docker registry).

Due to the external `registry.developer.com` domain being restricted only to the client's gateway, the HTTP challenge is out of question and we have to use the DNS challenge. But that hits [this issue](https://github.com/xenolf/lego/issues/570) instead. Basically – the lego library uses [this algorithm](https://github.com/xenolf/lego/blob/master/acme/dns_challenge.go#L235) to resolve the domains and for the external `registry.developer.com` domain properly to the relevant  Cloudflare DNS zone, but the internal `registry.int.developer.com` domain hits our internal DNS first, leading library to think there should be a `int.develop.com` zone in Cloudflare, which there isn't, causing the certificate provisioning to fail.

Lego allows overriding of the DNS servers used with a [`--dns-resolvers`](https://github.com/xenolf/lego/blob/master/cli_handlers.go#L42-L51) parameter which allows to force it to use a DNS server which will resolve the zone properly like, say, Google's 8.8.8.8 (I don't know if this is the correct way to solve this issue, but it works). Træfik does not expose that parameter in the relevant [configuration section](https://github.com/containous/traefik/blob/master/provider/acme/provider.go#L73-L76). I have tested and adding it works correctly for my use case (thought the way the DNS resolvers are a global variable in lego could cause race conditions if/when per-domain DNS challenge settings will be available with a naive implementation).

Would you be willing to expose such a parameter in configuration?

**Comment by trustfather** [comment]
I need it too.

**Comment by jaen** [comment]
@trustfather if you need it very badly feel free to try if my hacky addition of the configuration option here https://hub.docker.com/r/jaen/traefik-dns-resolvers works for you (code's here if you need to audit it: https://github.com/jaen/traefik/tree/feature/lets-encrypt-dns-challenge-dns-resolvers-configuration).

In my use case it seems to work for when I specify ACME config like so:
```toml
[acme.dnsChallenge]
provider = "cloudflare"
delayBeforeCheck = 60
dnsResolvers = [
  "8.8.8.8"
]
```

Can't vouch for correctness of it all though – config parsing seems rather intricate and DNS resolvers are a global variable in lego, so there me be some race condition corner case lurking somewhere in such a naive implementation.

**Comment by bitdegree67** [comment]
i need to specify DNS too for DNS challenge

**Comment by comnam90** [comment]
> @trustfather if you need it very badly feel free to try if my hacky addition of the configuration option here https://hub.docker.com/r/jaen/traefik-dns-resolvers works for you (code's here if you need to audit it: https://github.com/jaen/traefik/tree/feature/lets-encrypt-dns-challenge-dns-resolvers-configuration).
> 
> In my use case it seems to work for when I specify ACME config like so:
> 
> ```toml
> [acme.dnsChallenge]
> provider = "cloudflare"
> delayBeforeCheck = 60
> dnsResolvers = [
>   "8.8.8.8"
> ]
> ```
> 
> Can't vouch for correctness of it all though – config parsing seems rather intricate and DNS resolvers are a global variable in lego, so there me be some race condition corner case lurking somewhere in such a naive implementation.
> 
> EDIT: I swear there was a comment here I replied to last time I was looking at this

Any chance of you making a pull request to get this merged into the core code....?

**Comment by traefiker** [comment]
Closed by #4101.
